File: Tries/JS/Tries.js
class Tries {
    constructor(value = '') {
       this.children = new Map(); // <letter,new Map()>
       this.isWord = false; // to check if word or not
       this.value = value; // Value at current Tries
    }
 
    push(word) {
       let cur = this;
       for (let letter of word) { // traverse the given word
          if (!cur.children.has(letter)) cur.children.set(letter, new Tries(letter)); // if not present push it  
          cur = cur.children.get(letter); // Get to next map
       }
       cur.isWord = true; // True if word
    }
 
    checkWord(word) {
       let cur = this;
       for (let letter of word) {
          if (!cur.children.has(letter)) return false;
          cur = cur.children.get(letter);
       }
       return cur.isWord === true ? true : false;
 
    }
 
    checkPrefix(word) {
       let cur = this;
       for (let letter of word) {
          if (!cur.children.has(letter)) return false;
          cur = cur.children.get(letter);
       }
       return true;
 
    }
    /**
     * 
     * @param {String} prefix 
     * @returns last node of curr prefix
     */
    getLastPrexfix(prefix) {
       let cur = this;
       for (let letter of prefix) cur = (!cur.children.has(letter)) ? false : cur.children.get(letter);
       return cur;
    }
 
    /**
     * 
     * @param {string} letter 
     * @returns array of words
     */
    getPrefix(letter) {
       let words = [];
       let currNode = this.getLastPrexfix(letter); // get last node of prefix
       if (!currNode) return words; // if not present return false
       if (currNode.isWord) words.push(letter); // if the curr prefi is word add to words
       currNode.children.forEach((childs) => { // call of each child in childrens
          this.getAllWords(letter, childs, words);
       });
       return words;
    }
 
    /**
     * 
     * @param {string} prefix 
     * @param {Tries} childs 
     * @param {[]} words 
     * @returns []
     */
    getAllWords(prefix, childs, words) {
       if (!childs) return words;
       prefix += childs.value;
       if (childs.isWord) words.push(prefix);
       childs.children.forEach((c) => {
          this.getAllWords(prefix, c, words);
       });
       return words;
    }
 
    /**
     * 
     * @param {string} word 
     * @returns 
     */
    remove(word) {
       if (!word) return false;
       let curr = this;
       let stack = [];
 
       // add all node into stack correspond to words
       for (let x of word) {
          if (!curr.children.has(x)) return false;
          stack.push(curr);
          curr = curr.children.get(x);
       }
       curr.isWord = false;
       // remove curr's child having value prev.value
       while (stack.length > 1 && !curr.isWord) {
          let prev = curr;
          curr = stack.pop();
          if (prev.children.length > 0) break;
          curr.children.delete(prev.value);
       }
 
       return true;
    }
 };
 
 let t = new Tries();
 t.push("Abhi");
 t.push("Asshole");
 t.push("Apple");
 console.log(t.checkPrefix("As"));
 let c = t.getPrefix("A");
 console.log(c);
 t.remove("Abhi");
 t.remove("Apple");
 c = t.getPrefix("A");
 console.log(c);

File: Tries/CPP/Tries.cpp
#include <iostream>
using namespace std;

// Define a Trie class
class Trie {
private:
    Trie* child[26] = {}; // An array of pointers for each character ('a' to 'z')
    bool isWord = false; // Indicates if this node represents a complete word

    // Recursive function to insert a word into the Trie
    void insert2(string word, Trie* prev) {
        // If the word is empty, mark the current node as a complete word
        if (word.size() == 0) {
            prev->isWord = true;
            return;
        }

        Trie* curr = prev; // Current Trie node
        int idx = word[0] - 'a'; // Calculate the index for the first character
        if (curr->child[idx] == nullptr) curr->child[idx] = new Trie(); // If the child doesn't exist, create a new Trie node
        insert2(word.substr(1), curr->child[idx]); // Recursively insert the rest of the word
    }

    // Iterative function to insert a word into the Trie
    void insertChar(string word) {
        Trie* current = this; // Current Trie node
        for (auto ch : word) {
            int idx = ch - 'a'; // Calculate the index for the character
            if (current->child[idx] == nullptr)
                current->child[idx] = new Trie(); // If the child doesn't exist, create a new Trie node
            current = current->child[idx]; // Move to the next Trie node
        }
        current->isWord = true; // Mark the last node as a complete word
    }

public:
    // Public function to insert a word into the Trie
    void insert(string s) {
        // insertChar(s); // You can choose between recursive or iterative insert
        insert2(s, this); // This uses the recursive insert
    }

    // Public function to search for a complete word in the Trie
    bool searchWord(string word) {
        Trie* current = this; // Start at the root of the Trie
        for (auto ch : word) {
            int idx = ch - 'a'; // Calculate the index for the character
            if (current->child[idx] == nullptr)
                return false; // If the child doesn't exist, the word is not in the Trie
            current = current->child[idx]; // Move to the next Trie node
        }
        return current->isWord; // Check if the last node represents a complete word
    }

    // Public function to check if a prefix exists in the Trie
    bool prefixCheck(string word) {
        Trie* current = this; // Start at the root of the Trie
        for (auto ch : word) {
            int idx = ch - 'a'; // Calculate the index for the character
            if (current->child[idx] == nullptr)
                return false; // If the child doesn't exist, the prefix is not in the Trie
            current = current->child[idx]; // Move to the next Trie node
        }
        return true; // The prefix is found in the Trie
    }
};

int main() {
    Trie* t = new Trie(); // Create a new Trie instance
    t->insert("abhishek"); // Insert a word into the Trie
    cout << t->searchWord("abhi") << endl; // Search for a complete word
    cout << t->prefixCheck("ab") << endl; // Check if a prefix exists
    cout << t->prefixCheck("abhi") << endl; // Check if a prefix exists
    cout << t->prefixCheck("abhishek") << endl; // Check if a prefix exists

    return 0;
}
